### PR TITLE
Update tools

### PR DESCRIPTION
**Pull request by AJenbo**
Type: bugfix / refactoring
Breaking change: no

- [RandomException](https://www.php.net/manual/en/class.random-randomexception.php) is only thrown in 8.2+
- phpcs/php-cs-fixer: Now detects a few more redundant return types, if clause must start on first line.
- phpunit: Locked to 10.5.31 since 10.5.32 has a breaking change that conflicts with Paratest, see https://github.com/paratestphp/paratest/pull/883#issuecomment-2330971002 for alternatives.
- phpstan: stricter checks for preg-functions

### CODE REVIEW

File: src/main/php/PDepend/Application.php [hunk 1]
```diff
@@ -211,8 +211,7 @@ private function getAvailableOptionsFor(string $serviceTag): array
 
         foreach ($loggerServices as $loggerServiceTags) {
             foreach ($loggerServiceTags as $loggerServiceTag) {
-                if (
-                    isset($loggerServiceTag['option'], $loggerServiceTag['message'])
+                if (isset($loggerServiceTag['option'], $loggerServiceTag['message'])
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Application.php:214]: Annoying, this has conflict potential with open PRs but I don't find the new indentation any better.

**Review by kylekatarnls — APPROVED** [review]

**Comment by kylekatarnls** [comment]
Hello, I think it's overkill to run PHPStan in each PHPUnit, it's also a bit messy as a result (when it fails).

What do you think about migrating to:
image

Like in:
https://github.com/phpmd/phpmd/pull/1058/files

**Comment by AJenbo** [comment]
Not running it for each php version means we might miss compatibility issues (it analyzes things specifically for the executing version). This PR come from phpstan detecting an issue that only happens under php 8.1

We could split phpunit and phpstan but we woulg have a lot more jobs and a lot more overhead as a result.

**Comment by kylekatarnls** [comment]
> This PR come from phpstan detecting an issue that only happens under php 8.1

Where?

**Comment by AJenbo** [comment]
Here: https://github.com/phpmd/phpmd/actions/runs/10763472593/job/29845041568

**Comment by kylekatarnls** [comment]
Fishy, we should inspect why only 8.1 + stable, we should be sure it's not caused by installed version of phpstan, composer cache or something like this.

IMO, it still should be a separated task (which can run against multiple versions if relevant like we do for PHPUnit, and run with stable/lowest but honestly I don't see why it would make sense for PHPStan)

Even PHP version should barely give different result, I see very few case, there is `DateTime::modify()` signature moving from `DateTime|false` to `DateTime` in PHP 8.3 or things like that, but it sounds to me so few difference would not worth run all the version.

However I'm fine to continue to run all the version, but just not piggy-backing it on the PHPUnit job, BTW I would also split the codestyle job into separate jobs for each checker. And all of those job should be grouped in a "Codestyle" category rather than "Test" I think.

**Comment by AJenbo** [comment]
PHPStan 1.12.1 is not able to detect the issue, 1.12.2 can. I'm not sure exactly which change fixed this, but it's likely the `Run @mixin class reflection extensions after all other class reflection extensions` bugfix as it was causing part of extensions that where analyzed with extensions to be ignored.

Running each tool as a separate job causes a lot of overhead and the many jobs can cause them to get queued as well, especially when multiple PRs are being worked on simultaneously, or being frequently updated. I'm guessing your argument for splitting them is to not have to open the log to see which step failed?

We should at least run it for the oldest and latest support PHP. It's very helpful for catching deprecated things (e.g. 8.4 implicit nullable, count() not accepting null) when using it with the latest version, and for the oldest version we catch comparability issues like in this case.

Running it with min/latest dependencies likewise help us detect comparability issues and breaking changes that we need to handle.

**Comment by kylekatarnls** [comment]
> Running each tool as a separate job causes a lot of overhead and the many jobs can cause them to get queued as well

That's exactly the same if those are steps of a unique job, they will be queued. Also we run it 8 times (while even if we keep PHP version matrix, we likely don't need the stable/lowest, so we would go down to 4 times only, we'll end up with better parallelization and actually less total pipeline time)

> PHPStan 1.12.1 is not able to detect the issue, 1.12.2 can. 

So it's not about PHP version and running on all versions x stable+lowest would help to detect such problem only by luck as we're brute-forcing it.

> I'm guessing your argument for splitting them is to not have to open the log to see which step failed?

Yes, it's:
- to see quickly what is actually failing
- to see both (if test and PHPStan are both failing, at the moment, if PHPUnit fails it will hide the PHPStan result as it's not ran)
- standard, project commonly do static analyzer stuff in separate jobs
- consistent: no good reason to treat differently PHPStand, PHPCS and PHP-CS-fixer like we currently do

> We should at least run it for the oldest and latest support PHP.

And yes, nothing prevent you to use any matrix in each job as long as it makes sense, it's precisely the point of the matrix, we can make variables for what actually can make the job output different, and there is little chance that a matrix would just be good for all the tools at once.

**Comment by AJenbo** [comment]
> So it's not about PHP version and running on all versions x stable+lowest would help to detect such problem only by luck as we're brute-forcing it.

It _is_ about the PHP version. The class exists only from PHP 8.2 onwards. Detecting the issue requires a tool with the right feature set (PHPStan 1.12.2+). I wouldn't call that luck or brute force. 

> That's exactly the same if those are steps of a unique job

The environment setup and teardown takes about 21 seconds. When tasks share the same environment in a single job, we only incur that cost once. Splitting them into multiple jobs means paying that cost repeatedly, adding almost 3 minutes of extra CPU time, and that’s not counting the delay before jobs are picked up.

> We'll end up with better parallelization and actually less total pipeline time

PHPUnit takes 5 seconds, and PHPStan takes 20 seconds to run, so adding 21 seconds of overhead is significant. Waiting an additional 5 seconds in a serial flow can be better than ending up in queues. Splitting jobs also pushes us closer to GitHub’s concurrency limit of 20 jobs. Right now, we have 11 jobs; splitting them would bring us to 19. Testing on PHP 8.4+1 will push that to 23, leading to queue delays even for a single concurrent commit.

> we likely don't need the stable/lowest

Without stable/lowest testing, you won’t catch issues with older versions of Symfony (or other dependencies). A better approach might be testing stable/lowest dependencies with stable/lowest PHP versions. But if we question that, then we should also question why we run PHPUnit for every PHP version?

> to see both (if test and PHPStan are both failing, at the moment, if PHPUnit fails it will hide the PHPStan result as it's not ran)

That’s true, but we also cancel other matrix jobs if one fails. So, if PHPUnit fails for PHP 8.2, we may not see additional failures under PHP 8.1 either. Developers should ideally test failing environments before pushing again, unless they’re okay with multiple iterations. Also, PHPUnit and PHPStan often flag the same issues, so it might just add noise to see both errors before resolving one fully.

> no good reason to treat differently PHPStand, PHPCS and PHP-CS-fixer like we currently do

PHPStan is a static analyzer for identifying incorrectly handled cases and faults. Just like with PHPUnit, different APIs (dependencies or runtime) should be analyzed to catch edge cases. PHPCS and PHP-CS-fixer, on the other hand, enforce stylistic rules unrelated to runtime. If you group them together, you’d also need to include PHPUnit, which isn’t appropriate as they serve different roles.

**Comment by kylekatarnls** [comment]
Great explanation! 👍 At the moment, I'm convinced and OK to keep our `.github/workflow` folder as is.

**Pull request merged**